tools/autograd/gen_python_functions.py

```diff
@@ -1541,3 +886,115 @@ def op_name(declaration):
+# old codegen script (which operates on Declarations.yaml).
+
+# TODO: remove all these HACKs after migration is completed!
+
+# function schema str -> NativeFunction
+NF_TABLE: Optional[Dict[str, NativeFunction]] = None
+
+def init(native_yaml_path: str) -> None:
+    from tools.codegen.gen import parse_native_yaml
+    global NF_TABLE
+    NF_TABLE = {str(f.func): f for f in parse_native_yaml(native_yaml_path)}
+
+# Multiple decl entries can map to the same native function (because of deprecated decl).
+def decl_to_native_function(decl: Dict[str, Any]) -> NativeFunction:
+    assert NF_TABLE is not None, 'need to initialize codegen.api.python with init()'
+    function_schema_str = decl['schema_string'][6:]
+    assert function_schema_str in NF_TABLE, f'cannot find func: {function_schema_str}'
+    return NF_TABLE[function_schema_str]
+
+# Each decl entry has unique python signature.
+def decl_to_python_signature(decl: Dict[str, Any], *, method: bool) -> PythonSignature:
+    f = decl_to_native_function(decl)
+
+    @with_native_function
+    def go(f: NativeFunction) -> PythonSignature:
+        return signature(f, method=method)
+
+    python_sig = go(f)
+
+    if decl.get('deprecated', False):
+        # TODO: directly load 'deprecated.yaml'.
+        # deprecated.yaml doesn't have complete type information, we need
+        # leverage the source signature (to which it delegates the call).
+        # Deprecated signature might reorder input_args and input_kwargs,
+        # but never changes output_args nor python_binding_args (if any?),
+        # so here we only look into these two types of args.
+        src_args: Dict[str, PythonArgument] = {a.name: PythonArgument(
+            name=a.name,
+            type=a.type,
+            cpp_type_str=a.cpp_type_str,
+            default=None,
+            default_init=None,
+        ) for a in itertools.chain(python_sig.input_args, python_sig.input_kwargs)}
+        args: List[Dict[str, Any]] = decl['arguments']
+        input_arg_names: List[str] = \
+            list(str(a['name']) for a in args if not a['kwarg_only'] and not a['output'])
+        input_kwarg_names: List[str] = \
+            list(str(a['name']) for a in args if a['kwarg_only'] and not a['output'])
+        python_sig = PythonSignatureDeprecated(
+            name=python_sig.name,
+            input_args=tuple(src_args[n] for n in input_arg_names if not method or n != 'self'),
+            input_kwargs=tuple(src_args[n] for n in input_kwarg_names),
+            output_args=python_sig.output_args,
+            tensor_options_args=python_sig.tensor_options_args,
+            method=python_sig.method,
+            deprecated_args_names=tuple(str(a['name']) for a in args),
+            deprecated_args_exprs=tuple(decl.get('call_args')),
+        )
+    return python_sig
+
+
+def gen_lambda_args_str(f: NativeFunction, method: bool, ps: PythonSignature) -> str:
+
+    @with_native_function
```

Comment:
Is there a reason to switch between creating a local function here vs. putting the decorator on the outer function (like in `gen_lambda_return_str`)? The second looks more concise (similar comment for a few other functions further down)

Reply:
The simple answer is that "@with_native_function" annotation requires the function only takes 'NativeFunction' argument :) And '@with_native_function' is required by some new codegen APIs.

I guess it's the functional flavor the new codegen should stick with - everything is generated from list of native functions via a chain of lambdas...

For python codegen, the situation is a bit more complex because of deprecated.yaml: deprecated and non-deprecated python signatures might associate with the same native function. So I might need create a new '@with_python_signature' wrapper to setup necessary context/local for this codegen script.

As I commented above - the integration code in this file is subjected to change. I'll clean up after changing all logic in this file to use the new codegen api.

Reply:
Somewhat relatedly, I'm not sure these small local functions here are buying you very much. In the tools.codegen.gen each of the functions is doing a pretty decent chunk of work (e.g., all of the codegen for a particular file), but here these are very small functions doing only a very small slice of the work, and needing to be orchestrated altogether by the legacy code.

Reply:
Right, these are really because I was gradually migrating logic from emit_single_dispatch() to new codegen api, method by method.
I do plan to merge them into one generator, equivalent to what emit_single_dispatch() currently does.

Reply:
I've merged these methods into emit_single_dispatch.